test/pipelinefinally_test.go

```diff
@@ -0,0 +1,457 @@
+			}},
+		},
+	}
+	if _, err := c.TaskClient.Create(finalTask); err != nil {
+		t.Fatalf("Failed to create final Task: %s", err)
+	}
+
+	pipeline := &v1beta1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{Name: pipelineName, Namespace: namespace},
+		Spec: v1beta1.PipelineSpec{
+			Tasks: []v1beta1.PipelineTask{{
+				Name:    "dagtask1",
+				TaskRef: &v1beta1.TaskRef{Name: taskName},
+			}, {
+				Name:    "dagtask2",
+				TaskRef: &v1beta1.TaskRef{Name: delayedTaskName},
+			}},
+			Finally: []v1beta1.PipelineTask{{
+				Name:    "finaltask1",
+				TaskRef: &v1beta1.TaskRef{Name: finalTaskName},
+			}},
+		},
+	}
+	if _, err := c.PipelineClient.Create(pipeline); err != nil {
+		t.Fatalf("Failed to create Pipeline: %s", err)
+	}
+
+	pipelineRun := &v1beta1.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{Name: pipelineRunName, Namespace: namespace},
+		Spec: v1beta1.PipelineRunSpec{
+			PipelineRef: &v1beta1.PipelineRef{Name: pipelineName},
+		},
+	}
+	if _, err := c.PipelineRunClient.Create(pipelineRun); err != nil {
+		t.Fatalf("Failed to create Pipeline Run `%s`: %s", pipelineRunName, err)
+	}
+
+	if err := WaitForPipelineRunState(c, pipelineRunName, timeout, PipelineRunFailed(pipelineRunName), "PipelineRunFailed"); err != nil {
+		t.Fatalf("Waiting for PipelineRun %s to fail: %v", pipelineRunName, err)
+	}
+
+	taskrunList, err := c.TaskRunClient.List(metav1.ListOptions{LabelSelector: "tekton.dev/pipelineRun=" + pipelineRunName})
+	if err != nil {
+		t.Fatalf("Error listing TaskRuns for PipelineRun %s: %s", pipelineRunName, err)
+	}
+
+	var dagTask1Time, dagTask2Time, finalTaskTime *metav1.Time
+	// verify non final task failed and final task executed and succeeded
+	for _, taskrunItem := range taskrunList.Items {
+		switch n := taskrunItem.Name; {
+		case strings.HasPrefix(n, pipelineRunName+"-finaltask1"):
+			if err := WaitForTaskRunState(c, n, TaskRunSucceed(n), "TaskRunSuccess"); err != nil {
+				t.Errorf("Error waiting for TaskRun to succeed: %v", err)
+			}
+			finalTaskTime = taskrunItem.Status.StartTime
+		case strings.HasPrefix(n, pipelineRunName+"-dagtask1"):
+			if !isFailed(t, n, taskrunItem.Status.Conditions) {
+				t.Fatalf("TaskRun %s for non final task should have failed", n)
+			}
+			dagTask1Time = taskrunItem.Status.CompletionTime
+		case strings.HasPrefix(n, pipelineRunName+"-dagtask2"):
+			if err := WaitForTaskRunState(c, n, TaskRunSucceed(n), "TaskRunSuccess"); err != nil {
+				t.Errorf("Error waiting for TaskRun to succeed: %v", err)
+			}
+			dagTask2Time = taskrunItem.Status.CompletionTime
+		default:
+			t.Fatalf("TaskRuns were not found for both final and dag tasks")
+		}
+	}
+	// final task should start executing after dagtask1 fails and dagtask2 is done
+	if finalTaskTime.Before(dagTask1Time) || finalTaskTime.Before(dagTask2Time) {
+		t.Fatalf("Final Tasks should start getting executed after all DAG tasks finishes")
+	}
+}
+
+// Pipeline exits with success even if final task execution fails
```

Comment:
i think this comment might be out of date since the pipelienrun is failing as far as i can tell (in this case the comments above the functions might not be adding much to the function names, might be easier to remove them)